nautobot_data_validation_engine/custom_validators.py

```diff
@@ -118,6 +128,140 @@ def clean(self):
+            if is_data_compliance_rule(validator):
+                compliance_rulesets.setdefault(validator.model, [])
+                compliance_rulesets[validator.model].append(validator)
+
+    return compliance_rulesets
+
+
+def get_classes_from_git_repo(repo: GitRepository):
+    """Get list of DataComplianceRule classes found within the custom_validators folder of the given repo."""
+    ensure_git_repository(repo, head=repo.current_head)
+    class_list = []
+    for importer, discovered_module_name, _ in pkgutil.iter_modules([f"{repo.filesystem_path}/custom_validators"]):
+        if discovered_module_name in sys.modules:
+            del sys.modules[discovered_module_name]
+        module = importer.find_module(discovered_module_name).load_module(discovered_module_name)
+        for _, complance_class in inspect.getmembers(module, is_data_compliance_rule):
+            class_list.append(complance_class)
+    return class_list
+
+
+class ComplianceError(ValidationError):
+    """A compliance error is raised only when an object fails an compliance check."""
+
+
+class DataComplianceRule(CustomValidator):
+    """Class to handle a set of validation functions."""
+
+    class_name: Optional[str] = None
+    model: str
+    result_date: timezone
+    enforce = False
+
+    def __init__(self, obj):
+        """Initialize an DataComplianceRule object."""
+        super().__init__(obj)
+        self.class_name = self.class_name or self.__class__.__name__
+        self.result_date = timezone.now()
+
+    def audit(self):
+        """Not implemented.  Should raise an ComplianceError if an attribute is found to be invalid."""
+        raise NotImplementedError
+
+    def mark_existing_attributes_as_valid(self, exclude_attributes=None):
+        """Mark all existing fields (any that were previously created) as valid=True."""
+        instance = self.context["object"]
+        if not exclude_attributes:
+            exclude_attributes = []
+        attributes = (
+            DataCompliance.objects.filter(
+                compliance_class_name=self.class_name,
+                content_type=ContentType.objects.get_for_model(instance),
+                object_id=instance.id,
+            )
+            .exclude(validated_attribute__in=["all"] + exclude_attributes)
+            .values_list("validated_attribute", flat=True)
+        )
+        for attribute in attributes:
+            self.compliance_result(message=f"{attribute} is valid.", attribute=attribute)
+
+    def clean(self):
+        """Override the clean method to run the audit function."""
+        try:
+            self.audit()
+            self.mark_existing_attributes_as_valid()
+            self.compliance_result(message=f"{self.context['object']} is valid")
+        except ComplianceError as ex:
+            exclude_attributes = []
+            try:
+                for attribute, messages in ex.message_dict.items():
+                    exclude_attributes.append(attribute)
```

Comment:
Added some comments, but basically the idea is "I need a list of attributes that I'm currently using to create/update `DataCompliance` objects so that I can exclude them from the later `mark_existing_attributes_as_valid` call."

I can get the excluded attributes using `exclude_attributes = ex.message_dict.keys()`, but I'm still going to have to iterate over the items to create/update the `DataCompliance` objects.